File: aws-ec2/operation/describe_image.rb
# begin
@log.trace("Started executing 'fb-cloud:aws-ec2:operation:describe_image.rb' flintbit...")
begin
    # Flintbit Input Parameters
    # Mandatory
    connector_name = @input.get('connector_name')	# Name of the Amazon EC2 Connector
    action = 'describe-image'	# Contains the name of the operation: describe_image
    image_id = @input.get('image-id')	# Specifies the image ID of Amazon EC2

    # Optional
    @access_key = @input.get('access-key')
    @secret_key = @input.get('security-key')
    region = @input.get('region')	# Amazon EC2 region (default region is "us-east-1")
    request_timeout = @input.get('timeout')	# Execution time of the Flintbit in milliseconds (default timeout is 60000 milloseconds)

    @log.info("Flintbit input parameters are, action : #{action} | image_id : #{image_id} | region : #{region}")

    if connector_name.nil? || connector_name.empty?
        raise 'Please provide "Amazon EC2 connector name (connector_name)" to describe Image'
    end

    if image_id.nil? || image_id.empty?
        raise 'Please provide "Amazon EC2 image ID (image_id)" to describe image'
    end

    connector_call = @call.connector(connector_name).set('action', action).set('image-id', image_id).set('access-key', @access_key).set('security-key', @secret_key)

    if !region.nil? && !region.empty?
        connector_call.set('region', region)
    else
        @log.trace("region is not provided so using default region 'us-east-1'")
    end

    if request_timeout.nil? || request_timeout.is_a?(String)
        @log.trace("Calling #{connector_name} with default timeout...")
        response = connector_call.sync
    else
        @log.trace("Calling #{connector_name} with given timeout #{request_timeout}...")
        response = connector_call.timeout(request_timeout).sync
    end

    # Amazon EC2 Connector Response Meta Parameters
    response_exitcode = response.exitcode	# Exit status code
    response_message = response.message	# Execution status messages
    image_info = response.get('image-info')	# image info json

    if response_exitcode == 0
        @log.info("SUCCESS in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
        @log.info("Amazon EC2 Image info : #{image_info}")

        @output.set('exit-code', 0).set('message', response_message).set('image-info', image_info)
    else
        @log.error("ERROR in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
        response=response.to_s
        if !response.empty?
        @output.set('message', response_message).set('exit-code', 1).setraw('error-details',response)
        else
        @output.set('message', response_message).set('exit-code', 1)
        end
    end
rescue Exception => e
    @log.error(e.message)
    @output.set('exit-code', 1).set('message', e.message)
end
@log.trace("Finished executing 'fb-cloud:aws-ec2:operation:describe_image.rb' flintbit")
# end

File: aws-ec2/operation/workflow/describe_image.js
/**
 * Creation Date: 7/06/2019
 * Summary: To Describe Image on AWS
 * Description: To Describe Image on AWS EC2 using aws-ec2 connector
 */

log.trace("Started executing the flintbit 'fb-cloud:aws-ec2:operation:workflow:describe_image.js'");


input_scope = JSON.parse(input);   //For checking if the input contains the necessary keys

//Connector Name  - mandatory
connector_name = "amazon-ec2";
connector_call = call.connector(connector_name);
log.info("Connector Name: "+connector_name);

//Action - mandatory
action = "describe-image";
log.info("Action: "+action);

connector_call.set("action",action);

if(input_scope.hasOwnProperty("cloud_connection")){
    
    //Access Key - mandatory
    access_key = input_scope.cloud_connection.encryptedCredentials["access_key"];
    if(access_key!=null || access_key!=""){
       connector_call.set("access-key",access_key); 
       log.info("Access Key is given");
    }
    else{
        log.error("Access-Key is null or empty string.");
    }

    //Security Key - mandatory
    secret_key = input_scope.cloud_connection.encryptedCredentials["secret_key"];
    if(secret_key!=null || secret_key!=""){
       connector_call.set("security-key",secret_key); 
       log.info("Security Key is given");
    }
    else{
        log.error("Security-Key is null or empty string.");
    }

    //Image ID
    if(input_scope.hasOwnProperty("image_id")){
        image_id = input.get("image_id");
        if(image_id!=null || image_id!=""){
            connector_call.set("image-id",image_id); 
            log.info("Image ID: "+image_id);
        }
        else{
            log.error("Image ID is null or empty string.")
        }
    }
    else{  //image_id key not present in input JSON 
        log.error("Input does not contain the key 'image_id'")
    }

    //Region - mandatory
    if(input_scope.hasOwnProperty("region")){
        region = input.get("region");
        if(region!=null || region!=""){
            connector_call.set("region",region); 
            log.info("Region: "+region);
        }
        else{
            log.error("Region is null or empty string.");
        }
    }
    else{  //region key not present in input JSON 
        log.error("Input does not contain the key 'region'");
    }

    //Timeout - NOT mandatory
    if(input_scope.hasOwnProperty("request_timeout")){
        request_timeout = input.get("request_timeout");
        if(request_timeout!=null || request_timeout!=""){
            connector_call.set("timeout",request_timeout); 
            log.info("Request Timeout: "+request_timeout);
        }
        else{
            connector_call.set("timeout",240000); 
            log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
        }
    }
    else{
        connector_call.set("timeout",240000); 
        log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
    }

    //Connector Call
    response = connector_call.sync();

    //Response Metaparameters
    response_exitcode = response.exitcode();
    response_message = response.message();

    //Response Parameters
    image_info = response.get("image-info");

    if(response_exitcode==0){
        user_message = "<b>The Image Description is:</b><ul>";
        user_message = user_message + "    <li><b>Architecture:</b> "+image_info.get("architecture")+"</li>"+
                        "    <li><b>Description:</b> "+image_info.get("description")+"</li>"+
                        "    <li><b>Image Type:</b> "+image_info.get("image-type")+"</li>"+
                        "    <li><b>Image Name:</b> "+image_info.get("image-name")+"</li>"+
                        "    <li><b>Image Location:</b> "+image_info.get("image-location")+"</li>"+
                        "    <li><b>Kernel ID:</b> "+image_info.get("kernel-id")+"</li>"+
                        "    <li><b>Hypervisor:</b> "+image_info.get("hypervisor")+"</li></ul>";
        log.info(user_message);
        output.set("user_message",user_message)
            .set("exit-code",response_exitcode)
            .set("message",response_message)
            .set("result",image_info);
        log.trace("Finished executing the flintbit 'fb-cloud:aws-ec2:operation:workflow:describe_image.js' successfully");
    }
    else{
        log.error("Failure in execution, message:"+response_message+" | exitcode:"+response_exitcode);
        output.set("error",response_message).set("exit-code",-1);
        log.trace("finished executing 'fb-cloud:aws-ec2:operation:workflow:describe_image.js' with errors")
    }
}
else{   //cloud_connection key not present in input JSON
    log.error("Cloud Connection not given. Can not authenticate without Secret-Key and Access-Key");
}